cpp/src/join/mixed_join_kernels.cuh

```diff
@@ -0,0 +1,331 @@
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <cooperative_groups.h>
+
+#include <cub/cub.cuh>
+#include <cuco/detail/pair.cuh>
+#include <thrust/equal.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+
+namespace cudf {
+namespace detail {
+namespace cg = cooperative_groups;
+
+/**
+ * @brief Device functor to determine if two pairs are identical.
+ *
+ * This equality comparator is designed for use with cuco::static_multimap's
+ * pair* APIs, which will compare equality based on comparing (key, value)
+ * pairs. In the context of joins, these pairs are of the form
+ * (row_hash, row_id). A hash probe hit indicates that hash of a probe row's hash is
+ * equal to the hash of the hash of some row in the multimap, at which point we need an
+ * equality comparator that will check whether the contents of the rows are
+ * identical. This comparator does so by verifying key equality (i.e. that
+ * probe_row_hash == build_row_hash) and then using a row_equality_comparator
+ * to compare the contents of the row indices that are stored as the payload in
+ * the hash map.
+ *
+ * This particular comparator is a specialized version of the pair_equality used in hash joins. This
+ * version also checks the expression_evaluator.
+ */
+template <bool has_nulls>
+class pair_expression_equality {
+ public:
+  __device__ pair_expression_equality(
+    table_device_view build,
+    table_device_view probe,
+    bool swap_tables,
+    cudf::ast::detail::expression_evaluator<has_nulls> evaluator,
+    cudf::ast::detail::IntermediateDataType<has_nulls>* thread_intermediate_storage,
+    null_equality nulls_are_equal = null_equality::EQUAL)
+    : build(build),
+      probe{probe},
+      swap_tables{swap_tables},
+      nulls_are_equal{nulls_are_equal},
+      evaluator{evaluator},
+      thread_intermediate_storage{thread_intermediate_storage}
+  {
+  }
+
+  // The parameters are build/probe rather than left/right because the operator
+  // is called by cuco's kernels with parameters in this order (note that this
+  // is an implementation detail that we should eventually stop relying on by
+  // defining operators with suitable heterogeneous typing). Rather than
+  // converting to left/right semantics, we can operate directly on build/probe
+  // until we get to the expression evaluator, which needs to convert back to
+  // left/right semantics because the conditional expression need not be
+  // commutative.
+  __device__ __forceinline__ bool operator()(const pair_type& build_row,
+                                             const pair_type& probe_row) const noexcept
+  {
+    // TODO: I've inlined the logic from cudf's row_equality_comparator because
+    // that comparator's constructor is not visible on device, and changing it
+    // would require removing an assertion-raising test. I don't think we want
+    // to do that, but should verify before finalizing.
```

Comment:
Instead of needing to construct the `row_equality_comparator` on device, can we construct it on host and then pass it by value to the kernel? Then we can avoid needing to pass the `nulls_are_equal` parameter through the kernel parameters by just bundling it into the equality comparator passed to the kernel.

Reply:
We can't construct the `pair_expression_equality` comparator on the host because it needs access to the shared memory buffer. However, we could construct the `row_equality_comparator` on the host and pass it to the constructor of the `pair_expression_equality`. Would you prefer that?